akka-stream/src/main/scala/akka/stream/impl/fusing/Interpreter.scala

```diff
@@ -282,18 +339,62 @@ private[akka] class OneBoundedInterpreter(ops: Seq[Stage[_, _]], val forkLimit:
       null
     }
 
-    override def hold(): FreeDirective = {
-      if (currentOp.holding) throw new IllegalStateException("Cannot hold while already holding")
-      currentOp.holding = true
+    override def holdUpstream(): UpstreamDirective = {
+      currentOp.precedingWasPull = false
+      currentOp.holdingUpstream = true
+      exit()
+    }
+
+    override def holdUpstreamAndPush(elem: Any): UpstreamDirective = {
+      ReactiveStreamsCompliance.requireNonNullElement(elem)
+      if (comingFromUpstream && !currentOp.holdingDownstream)
+        throw new IllegalStateException("Cannot holdUpstreamAndPush unless isHoldingDownstream")
+      if (comingFromDownstream)
+        throw new IllegalStateException("Cannot holdUpstreamAndPush during onPull")
+      if (comingFromSide)
+        throw new IllegalStateException("Cannot holdUpstreamAndPush during onAsyncInput")
+      currentOp.precedingWasPull = false
+      currentOp.holdingUpstream = true
+      currentOp.holdingDownstream = false
+      elementInFlight = elem
+      state = Pushing
+      null
+    }
+
+    override def isHoldingUpstream: Boolean = currentOp.holdingUpstream
+
+    override def holdDownstream(): DownstreamDirective = {
+      currentOp.holdingDownstream = true
       exit()
     }
 
-    override def isHolding: Boolean = currentOp.holding
+    override def holdDownstreamAndPull(): DownstreamDirective = {
+      if (comingFromUpstream)
```

Comment:
I have an idea here. What if we encode the holding states as a bitfield instead of separate booleans, say holdingBalls, the current direction also as a bitfield (with exactly one bit set), say directionBall, and the reqirements of the directive as directiveNeedsBalls. Then we can use this:

    if ((holdingBalls | directionBall) & directiveNeedsBalls == directiveNeedsBalls)
      // do the action
    else
      calculateNiceException(holdingBalls, directionBall, directiveNeedsBalls)

That way the calculation looks simple and needs less branches, while the error message generation can be centralized and we do not care if it is fast or not.